### DISCUSSION TITLE
Does hardware-instanced drawing disanable early-z optimization?

### DISCUSSION THREAD

**Discussion by NETools** [MonoGame/MonoGame#8170] [Graphics]
Hello,

i am writing a batch-oriented occlusion culling query where several models can be added of which every model's component-wise min and max values are calculated which is then used to draw a cube containing the model. During the drawing process of the cube a writable structured buffer is addressed in the pixel shader which sets the visibility value of the corresponding index associated with that drawn cube (which corresponds with the model's id) to one indicating that the pixel shader stage was called for this cube (implying it is not occluded). But when using hardware instancing the pixel shader is called no matter whether or not the cube is occluded by another cube. Shouldnt at least one cube be discarded since the depth value of either of them is smaller (early-z optimization)?

Here some images to ilustrate the problem:
 
Both cubes are visible:
![image](https://github.com/MonoGame/MonoGame/assets/117414781/bc363217-1900-40f4-a292-7198276112b5)

Only one cube is visible (other cube is behind the camera):
![image](https://github.com/MonoGame/MonoGame/assets/117414781/b9ac1cd3-2699-4e29-8cdd-f8daa393b30b)

Only one cube is visible, the other occluded:
![image](https://github.com/MonoGame/MonoGame/assets/117414781/0a82c0fe-2c26-4933-a67a-6ee464483402)

The same situation with the green cube being occluded by the red one:
![image](https://github.com/MonoGame/MonoGame/assets/117414781/3406faaf-a1e3-4dad-88d0-dfcfde6b8e08)

Doesnt make sense to me.

**Comment by ThomasFOG** [comment]
To put it simply, depth testing happens _after_ the pixel shader by default, which means that both of the cubes get both drawn before any depth is being tested. And that's pretty much all there is to know.

Now to the technical unintuitive part: it may, or may not happen, that the GPU kind of know that a pixel might be useless to process. But that's up to the GPU to decide and this is something not quite documented and specific to each GPU. There are ways to kind of prepare things in a way that this had the best chances to happen, but the short answer is that you can't know and certainly not control it. You can't "disable early-z", it is the GPU that decides that for you (and each GPU may have its own sauce).

In your case, what seems to be happening, is that working with instanced drawing puts your GPU in a state where the depth pass is always after the pixel shader (which makes sense because each instance is likely processed in parallel, so the GPU _has_ to wait for all of them before it can know how to depth test them).

So you'll likely need to do your own sauce to optimize this if you really need to. And you can't assume that if your GPU does an early-z, that another GPU will do the same.
